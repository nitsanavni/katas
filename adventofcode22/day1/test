source <(jaq '.[]' tests.json | jaq -rf test-command.jq)
