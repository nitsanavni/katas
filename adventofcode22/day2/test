#!/usr/bin/bash

echo "Rock Paper Scissors"
echo "Part 1"
echo ""

echo "total round score"
jaq -n '{ round_outcome_score: 7, shape_score: 3}' | \
jaq -cr "$(cat rock-paper-scissors.jq) \"input:\", ., \"total score:\", total_round_score"
echo ""

echo "round outcome score"
jaq -n '"lost", "draw", "won" | { round_outcome: . }' | \
jaq -cr "$(cat rock-paper-scissors.jq) [.round_outcome, \" -> \", round_outcome_score | tostring] | add"
echo ""

echo "product"
jaq -n '["rock", "paper", "scissors"]|[.,.]' | \
jaq -cf ../../product-jq/product.jq 
echo ""

echo "round outcome"
jaq -n '["rock", "paper", "scissors"]|[.,.]' | \
jaq -f ../../product-jq/product.jq | \
jaq '{ ours: .[0], theirs: .[1] }' | \
jaq -cr "$(cat rock-paper-scissors.jq) [. | tostring, \" -> \", round_outcome] | add"
echo ""

input_combinations=$(jaq -n '[["A","B","C"],["X","Y","Z"]]' | jaq -f ../../product-jq/product.jq | jaq 'join(" ")')

echo "parse one round"
echo $input_combinations | \
jaq -cr "$(cat rock-paper-scissors.jq) [., \" -> \", parse | tostring] | add"
echo ""

echo "leap forward: round score"
echo $input_combinations | \
jaq -cr "$(cat rock-paper-scissors.jq) [., \" -> \", round_score | tostring] | add"
echo ""

echo "sum them up"
echo $input_combinations | \
jaq -r '.' | \
jaq -sRcr "$(cat rock-paper-scissors.jq) [., \" -> \", sum_them_up | tostring] | add"
echo ""

echo "with our input"
cat input.txt | \
jaq -sRcr "$(cat rock-paper-scissors.jq) [sum_them_up | tostring] | add"
echo ""

echo "Part 2"
echo ""

echo "parse 2"
echo $input_combinations | \
jaq -cr "$(cat rock-paper-scissors.jq) [., \" -> \", parse2 | tostring] | add"
echo ""

echo "choose shape"
echo $input_combinations | \
jaq -cr "$(cat rock-paper-scissors.jq) [(parse2 | tostring), \" -> \", (parse2 | choose_shape | tostring)] | add"
echo ""

echo "round score"
echo $input_combinations | \
jaq -cr "$(cat rock-paper-scissors.jq) [., \" -> \", round_score_part_2 | tostring] | add"
echo ""

echo "total"
echo $input_combinations | \
jaq -r '.' | \
jaq -sRrc "$(cat rock-paper-scissors.jq) [., \" -> \", sum_2 | tostring] | add"
echo ""

echo "with our input"
cat input.txt | \
jaq -sRcr "$(cat rock-paper-scissors.jq) [sum_2 | tostring] | add"
echo ""
