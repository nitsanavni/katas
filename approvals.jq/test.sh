#!/bin/bash

function find_test_files() {
    find . -name 'test_*.jq' | sort
}

function test_methods_in_file() {
    local file=$1
    grep -o 'def test_[^:^\(]*' $file | sed 's/def //' | sort
}

function run_test_method() {
    local module=$1
    local test_method=$2
    jq -nr "include \"$module\"; $test_method"
}

function verify_test_method() {
    local module=$1
    local test_method=$2
    local base="$module.$test_method"
    echo "  $test_method" >&2
    local received=$(run_test_method $module $test_method)
    local approved_file="$base.approved"
    local received_file="$base.received"
    echo "$received" >$received_file
    if [ ! -f $approved_file ]; then
        echo "  ðŸš« failed" >&2
        echo "  $approved_file not found" >&2
        echo "  received output is:" >&2
        cat $received_file >&2
        exit 1
    else
        if ! diff -u $approved_file $received_file; then
            echo "  ðŸš« failed" >&2
            exit 1
        else
            echo "  âœ… passed" >&2
            rm $received_file
        fi
    fi
}

function run_tests_in_file() {
    local file=$1
    local module=$(echo $file | sed 's/\.jq//' | sed 's/^\.\///')
    echo "" >&2
    echo $module >&2
    echo "" >&2
    local test_methods=$(test_methods_in_file $file)
    for test_method in $test_methods; do
        verify_test_method $module $test_method
        echo "" >&2
    done
    echo "" >&2
}

function run_tests() {
    local files=$(find_test_files)
    for file in $files; do
        run_tests_in_file $file
    done
}

run_tests
