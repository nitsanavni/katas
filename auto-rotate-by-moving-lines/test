./test.sh | verify -t rotate