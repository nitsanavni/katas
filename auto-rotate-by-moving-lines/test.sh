#!/usr/bin/bash

lib=$(cat lib.jq)

cat_input() {
cat << EOF
$input
EOF
}

echo test: rm typing indication

input="Bob
-- typing -->
Alice"
cat_input
echo -----
cat_input | jaq -sRr 'split("\n")[] | select(test("typing")|not)'
echo ""


echo test: swap lines

input="a
b
c"
cat_input
echo -----
echo 'swap(1;2)'
cat_input | jaq -sRr "$lib"'split("\n") | swap(1;2) | .[]'
echo 'swap(0;2)'
cat_input | jaq -sRr "$lib"'split("\n") | swap(0;2) | .[]'
echo ""


echo test: jq filter as bash var

filter="range(5)
| . * .
| . + 1"
jaq -nr "$filter"
echo ""
