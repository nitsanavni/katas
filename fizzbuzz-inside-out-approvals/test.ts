type Test = (name: string, printer: () => any) => void;

export const test: Test = (name, printer) => {
    console.log(`${name}`);
    console.log(`${printer()}`);
    console.log("");
};

export default test;
