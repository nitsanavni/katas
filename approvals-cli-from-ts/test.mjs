import test from "ava";
import { $, path } from "zx";
import { fileURLToPath } from "url";

test("hello world!", async (t) => {
  t.is(
    await $`python ../approvals-cli/verify.py  -t hello -r "hello world!"`
      .exitCode,
    0
  );
});
