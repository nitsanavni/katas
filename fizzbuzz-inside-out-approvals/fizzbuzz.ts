import product from "../product/product";
import test from "./test";

// "Fizz" | ""
// ["Fizz" | "", "Buzz" | ""]
// "" | "Fizz" | "Buzz" | "FizzBuzz"
// "" | "Fizz" | "Buzz" | "FizzBuzz" | number

const s = JSON.stringify;

const multipleOf: (d: number) => (n: number) => boolean = (d) => (n) =>
    n % d == 0;

test("multipleOf", () => {
    const ds = [3, 4];
    const ns = [3, 4, 8];

    return product(
        ds.map((d) => [d, multipleOf(d)]),
        ns
    )
        .map(([[d, ofFn], n]) => `of(${d})(${n}): ${ofFn(n)}`)
        .join("\n");
});

type StringLiteral<T> = T extends string
    ? string extends T
        ? never
        : T
    : never;

const code: <C>(a: {
    d: number;
    code: StringLiteral<C>;
}) => (n: number) => StringLiteral<C | ""> = ({ d, code }) => {
    const ofD = multipleOf(d);

    return (n: number) => (ofD(n) ? code : "");
};

test('"Fizz" | "" (code)', () => {
    const ns = [4, 5, 8, 9, 12];
    const codes = [
        { d: 4, code: "foo" },
        { d: 3, code: "fizz" },
    ];

    return product(codes, ns)
        .map(
            ([c, n]) =>
                `code({d:${c.d},code:"${c.code}"})(${n}): "${code(c)(n)}"`
        )
        .join("\n");
});

type Spec<C> = { d: number; code: StringLiteral<C> };
const arr: <C>(s: Spec<C>[]) => (n: number) => string[] = (s) => {
    const codes = s.map(code);

    return (n: number) => codes.map((c) => c(n));
};

const ns = [2, 3, 4, 12];
const specss = [
    [{ d: 3, code: "fizz" }],
    [{ d: 4, code: "foo" }],
    [
        { d: 3, code: "fizz" },
        { d: 4, code: "foo" },
    ],
    [
        { d: 4, code: "foo" },
        { d: 3, code: "fizz" },
    ],
];

test('["Fizz" | "", "Buzz" | ""]', () => {
    return product(specss, ns)
        .map(([specs, n]) => `arr(${s(specs)})(${n}):\n${s(arr(specs)(n))}`)
        .join("\n");
});

const codes: <C>(s: Spec<C>[]) => (n: number) => string = (s) => (n) =>
    arr(s)(n).join("");

test('"" | "Fizz" | "Buzz" | "FizzBuzz"', () => {
    return product(specss, ns)
        .map(([specs, n]) => `codes(${s(specs)})(${n}):\n${s(codes(specs)(n))}`)
        .join("\n");
});

const game: <C>(s: Spec<C>[]) => (n: number) => string | number =
    (s) => (n) => {
        const c = codes(s)(n);

        return c.length == 0 ? n : c;
    };

test('"" | "Fizz" | "Buzz" | "FizzBuzz" | number', () => {
    return product(specss, ns)
        .map(([specs, n]) => `game(${s(specs)})(${n}):\n${s(game(specs)(n))}`)
        .join("\n");
});

const fizzbuzz = game([
    { d: 3, code: "Fizz" },
    { d: 5, code: "Buzz" },
    { d: 7, code: "Whizz" },
    { d: 11, code: "Bang" },
]);

const range = (n: number): number[] => {
    const ret: number[] = [];

    for (let i = 0; i < n; i++) {
        ret.push(i);
    }

    return ret;
};

test("fizzbuzz", () => {
    const ns = range(100).map((x) => x + 1);

    return ns.map((n) => `${n}: ${fizzbuzz(n)}`).join("\n");
});
