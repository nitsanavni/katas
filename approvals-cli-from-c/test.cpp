#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"
#include <iostream> // std::cout
#include <sstream>  // std::stringstream
#include <string>   // std::string

extern "C" {
#include "fizzbuzz.h"
#include "verify.h"
}

const char *fizzbuzz_str() {
  std::stringstream ss;

  for (int i = 0; i < 100; i++) {
    ss << i << ": " << fizzbuzz(i) << "\n";
  }

  const std::string &str = ss.str();
  return strdup(str.c_str());
}

TEST_CASE("approvals from C") {
  CHECK(verify("appr from C", fizzbuzz_str()) == 0);
}
