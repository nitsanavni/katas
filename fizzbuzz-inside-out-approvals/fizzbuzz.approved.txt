multipleOf
of(3)(3): true
of(3)(4): false
of(3)(8): false
of(4)(3): false
of(4)(4): true
of(4)(8): true

"Fizz" | "" (code)
code({d:4,code:"foo"})(4): "foo"
code({d:4,code:"foo"})(5): ""
code({d:4,code:"foo"})(8): "foo"
code({d:4,code:"foo"})(9): ""
code({d:4,code:"foo"})(12): "foo"
code({d:3,code:"fizz"})(4): ""
code({d:3,code:"fizz"})(5): ""
code({d:3,code:"fizz"})(8): ""
code({d:3,code:"fizz"})(9): "fizz"
code({d:3,code:"fizz"})(12): "fizz"

["Fizz" | "", "Buzz" | ""]
arr([{"d":3,"code":"fizz"}])(2):
[""]
arr([{"d":3,"code":"fizz"}])(3):
["fizz"]
arr([{"d":3,"code":"fizz"}])(4):
[""]
arr([{"d":3,"code":"fizz"}])(12):
["fizz"]
arr([{"d":4,"code":"foo"}])(2):
[""]
arr([{"d":4,"code":"foo"}])(3):
[""]
arr([{"d":4,"code":"foo"}])(4):
["foo"]
arr([{"d":4,"code":"foo"}])(12):
["foo"]
arr([{"d":3,"code":"fizz"},{"d":4,"code":"foo"}])(2):
["",""]
arr([{"d":3,"code":"fizz"},{"d":4,"code":"foo"}])(3):
["fizz",""]
arr([{"d":3,"code":"fizz"},{"d":4,"code":"foo"}])(4):
["","foo"]
arr([{"d":3,"code":"fizz"},{"d":4,"code":"foo"}])(12):
["fizz","foo"]
arr([{"d":4,"code":"foo"},{"d":3,"code":"fizz"}])(2):
["",""]
arr([{"d":4,"code":"foo"},{"d":3,"code":"fizz"}])(3):
["","fizz"]
arr([{"d":4,"code":"foo"},{"d":3,"code":"fizz"}])(4):
["foo",""]
arr([{"d":4,"code":"foo"},{"d":3,"code":"fizz"}])(12):
["foo","fizz"]

"" | "Fizz" | "Buzz" | "FizzBuzz"
codes([{"d":3,"code":"fizz"}])(2):
""
codes([{"d":3,"code":"fizz"}])(3):
"fizz"
codes([{"d":3,"code":"fizz"}])(4):
""
codes([{"d":3,"code":"fizz"}])(12):
"fizz"
codes([{"d":4,"code":"foo"}])(2):
""
codes([{"d":4,"code":"foo"}])(3):
""
codes([{"d":4,"code":"foo"}])(4):
"foo"
codes([{"d":4,"code":"foo"}])(12):
"foo"
codes([{"d":3,"code":"fizz"},{"d":4,"code":"foo"}])(2):
""
codes([{"d":3,"code":"fizz"},{"d":4,"code":"foo"}])(3):
"fizz"
codes([{"d":3,"code":"fizz"},{"d":4,"code":"foo"}])(4):
"foo"
codes([{"d":3,"code":"fizz"},{"d":4,"code":"foo"}])(12):
"fizzfoo"
codes([{"d":4,"code":"foo"},{"d":3,"code":"fizz"}])(2):
""
codes([{"d":4,"code":"foo"},{"d":3,"code":"fizz"}])(3):
"fizz"
codes([{"d":4,"code":"foo"},{"d":3,"code":"fizz"}])(4):
"foo"
codes([{"d":4,"code":"foo"},{"d":3,"code":"fizz"}])(12):
"foofizz"

"" | "Fizz" | "Buzz" | "FizzBuzz" | number
game([{"d":3,"code":"fizz"}])(2):
2
game([{"d":3,"code":"fizz"}])(3):
"fizz"
game([{"d":3,"code":"fizz"}])(4):
4
game([{"d":3,"code":"fizz"}])(12):
"fizz"
game([{"d":4,"code":"foo"}])(2):
2
game([{"d":4,"code":"foo"}])(3):
3
game([{"d":4,"code":"foo"}])(4):
"foo"
game([{"d":4,"code":"foo"}])(12):
"foo"
game([{"d":3,"code":"fizz"},{"d":4,"code":"foo"}])(2):
2
game([{"d":3,"code":"fizz"},{"d":4,"code":"foo"}])(3):
"fizz"
game([{"d":3,"code":"fizz"},{"d":4,"code":"foo"}])(4):
"foo"
game([{"d":3,"code":"fizz"},{"d":4,"code":"foo"}])(12):
"fizzfoo"
game([{"d":4,"code":"foo"},{"d":3,"code":"fizz"}])(2):
2
game([{"d":4,"code":"foo"},{"d":3,"code":"fizz"}])(3):
"fizz"
game([{"d":4,"code":"foo"},{"d":3,"code":"fizz"}])(4):
"foo"
game([{"d":4,"code":"foo"},{"d":3,"code":"fizz"}])(12):
"foofizz"

fizzbuzz
1: 1
2: 2
3: Fizz
4: 4
5: Buzz
6: Fizz
7: Whizz
8: 8
9: Fizz
10: Buzz
11: Bang
12: Fizz
13: 13
14: Whizz
15: FizzBuzz
16: 16
17: 17
18: Fizz
19: 19
20: Buzz
21: FizzWhizz
22: Bang
23: 23
24: Fizz
25: Buzz
26: 26
27: Fizz
28: Whizz
29: 29
30: FizzBuzz
31: 31
32: 32
33: FizzBang
34: 34
35: BuzzWhizz
36: Fizz
37: 37
38: 38
39: Fizz
40: Buzz
41: 41
42: FizzWhizz
43: 43
44: Bang
45: FizzBuzz
46: 46
47: 47
48: Fizz
49: Whizz
50: Buzz
51: Fizz
52: 52
53: 53
54: Fizz
55: BuzzBang
56: Whizz
57: Fizz
58: 58
59: 59
60: FizzBuzz
61: 61
62: 62
63: FizzWhizz
64: 64
65: Buzz
66: FizzBang
67: 67
68: 68
69: Fizz
70: BuzzWhizz
71: 71
72: Fizz
73: 73
74: 74
75: FizzBuzz
76: 76
77: WhizzBang
78: Fizz
79: 79
80: Buzz
81: Fizz
82: 82
83: 83
84: FizzWhizz
85: Buzz
86: 86
87: Fizz
88: Bang
89: 89
90: FizzBuzz
91: Whizz
92: 92
93: Fizz
94: 94
95: Buzz
96: Fizz
97: 97
98: Whizz
99: FizzBang
100: Buzz

